File: coinchange2_bruteforce.py
# // Time Complexity :O(2^n)
# // Space Complexity :O(2^n)
class Solution:
    def __init__(self):
        self.count=0
    def change(self, amount: int, coins: List[int]) -> int:
        self.helper(amount,coins,0)
        return self.count
    def helper(self,amount,coins,index):
        
        #base
        if amount==0:
            self.count+=1
            return
        if amount<0 or index==len(coins):
            return
        #logic
        
        #choose
        self.helper(amount-coins[index],coins,index)
        
        #not choose
        self.helper(amount,coins,index+1)
        
        

File: CoinChange2BruteForce.java
// Brute force: 2^m*n (where m is the coins and n is the amount)
// Space complexity: 2^m*n

// Approach: brute force. recursion. Exhaustive approach

class Solution {
    int total = 0;
    public int change(int amount, int[] coins) {
        helper(amount, 0, coins, 0);
        return total;
    }
    
    private void helper(int target, int currentAmount, int[] coins, int idx) {
        // base
        if (currentAmount == target) {
            total++;
            return;
        }

        if (currentAmount > target || idx == coins.length) return;
        
        // logic
        // not choose
        helper(target, currentAmount, coins, idx+1);
        
        // choose
        currentAmount += coins[idx];
        helper(target, currentAmount, coins, idx);
    }
}